src/plugins/bsodmon/bsodmon.cpp

```diff
@@ -109,6 +109,113 @@
 #include "bugcheck.h"
 #include "plugins/output_format.h"
 
+static std::string to_hex_string(uint64_t value)
+{
+    std::stringstream stream;
+    stream << std::hex << value;
+    return stream.str();
+}
+
+static void save_file_metadata(
+    drakvuf_t drakvuf,
+    drakvuf_trap_info_t* info,
+    bsodmon* f,
+    uint64_t code,
+    uint64_t param1,
+    uint64_t param2,
+    uint64_t param3,
+    uint64_t param4)
+{
+    // TODO Add option
+    auto file = f->crashdump_dir + "/crashdump.metadata";
+    if (file.empty())
+        return;
+
+    FILE* fp = fopen(file.data(), "w");
+    if (!fp)
+        return;
+
+    json_object* jobj = json_object_new_object();
+    if (!jobj)
+    {
+        fclose(fp);
+        return;
+    }
+
+    json_object_object_add(jobj,
+        "KernelBase",
+        json_object_new_string(to_hex_string(drakvuf_get_kernel_base(drakvuf)).data()));
+
+    json_object_object_add(jobj,
+        "BugCheckCode",
+        json_object_new_string(to_hex_string(code).data()));
+
+    json_object_object_add(jobj,
+        "Param1",
+        json_object_new_string(to_hex_string(param1).data()));
+
+    json_object_object_add(jobj,
+        "Param2",
+        json_object_new_string(to_hex_string(param2).data()));
+
+    json_object_object_add(jobj,
+        "Param3",
+        json_object_new_string(to_hex_string(param3).data()));
+
+    json_object_object_add(jobj,
+        "Param4",
+        json_object_new_string(to_hex_string(param4).data()));
+
+    json_object_object_add(jobj,
+        "CR3",
+        json_object_new_string(to_hex_string(info->regs->cr3).data()));
+
+    json_object_object_add(jobj,
+        "RSP",
+        json_object_new_string(to_hex_string(info->regs->rsp).data()));
+
+    fprintf(fp, "%s\n", json_object_get_string(jobj));
+    fclose(fp);
+
+    json_object_put(jobj);
+}
+
+static void dump_system(
+    drakvuf_t drakvuf,
+    drakvuf_trap_info_t* info,
+    bsodmon* f,
+    uint64_t code,
+    uint64_t param1,
+    uint64_t param2,
+    uint64_t param3,
+    uint64_t param4)
+{
+    save_file_metadata(drakvuf, info, f, code, param1, param2, param3, param4);
+
+    auto file = f->crashdump_dir + "/crashdump.bin";
+    if (file.empty())
+        return;
+
+    FILE* fp = fopen(file.data(), "w");
```

Comment:
`FILE* fp= creat(file.data(), S_IWUSR | S_IRUSR);`

Reply:
We should not use `creat`:
* From https://linux.die.net/man/3/creat :
```
The creat() function is redundant. Its services are also provided by the open() function. It has been included primarily for historical purposes since many existing applications depend on it.
```
* From https://stackoverflow.com/a/33938315 :
```


fopen vs open in C

1) fopen is a library function while open is a system call.

2) fopen provides buffered IO which is faster compare to open which is non buffered.

3) fopen is portable while open not portable (open is environment specific).

4) fopen returns a pointer to a FILE structure(FILE *); open returns an integer that identifies the file.

5) A FILE * gives you the ability to use fscanf and other stdio functions.
```

Reply:
I have used `umask` to set permissions.